src/utils/string_parsing.ts

```diff
@@ -0,0 +1,295 @@
+ */
+
+import assert from "./assert";
+
+/**
+ * Convert a string to an Uint8Array containing the corresponding UTF-16 code
+ * units in little-endian.
+ * @param {string} str
+ * @returns {Uint8Array}
+ */
+function strToLeUtf16(str: string): Uint8Array {
+  const buffer = new ArrayBuffer(str.length * 2);
+  const res = new Uint8Array(buffer);
+  for (let i = 0; i < res.length; i += 2) {
+    const value = str.charCodeAt(i / 2);
+    res[i] = value & 0xFF;
+    res[i + 1] = value >> 8 & 0xFF;
+  }
+  return res;
+}
+
+/**
+ * Construct string from the little-endian UTF-16 code units given.
+ * @param {Uint8Array} bytes
+ * @returns {string}
+ */
+function leUtf16ToStr(bytes : Uint8Array) : string {
+  let str = "";
+  for (let i = 0; i < bytes.length; i += 2) {
+    str += String.fromCharCode((bytes[i + 1] << 8) + bytes[i]);
+  }
+  return str;
+}
+
+/**
+ * Convert a string to an Uint8Array containing the corresponding UTF-8 code
+ * units.
+ * @param {string} str
+ * @returns {Uint8Array}
+ */
+function strToUtf8(str : string) : Uint8Array {
+  // http://stackoverflow.com/a/13691499 provides an ugly but functional solution.
+  // (Note you have to dig deeper to understand it but I have more faith in
+  // stackoverflow not going down in the future so I leave that link.)
+
+  // Briefly said, `utf8Str` will contain a version of `str` where every
+  // non-ASCII characters will be replaced by an escape sequence of the
+  // corresponding representation of those characters in UTF-8.
+  // It does sound weird and unnecessarily complicated, but it works!
+  //
+  // Here is actually what happens with more words. We will rely on two browser
+  // APIs:
+  //
+  //   - `encodeURIComponent` will take a string and convert the non-ASCII
+  //     characters in it into the percent-encoded version of the corresponding
+  //     UTF-8 bytes
+  //     Example: encodeURIComponent("é") => 0xC3 0xA9 => `"%C3%A9"`
+  //
+  //   - `unescape` unescapes (so far so good) a percent-encoded string. But it
+  //     does it in a really simple way: percent-encoded byte by percent-encoded
+  //     byte into the corresponding extended ASCII representation on 8 bits.
+  //     As a result, we end-up with a string which actually contains instead of
+  //     each of its original characters, the UTF-8 code units (8 bits) of
+  //     those characters.
+  //     Let's take our previous `"é" => "%C3%A9"` example. Here we would get:
+  //     unecape("%C3%A9") => "\u00c3\u00a9" === "Ã©" (in extended ASCII)
+  //
+  // By iterating on the resulting string, we will then be able to generate a
+  // Uint8Array containing the UTF-8 representation of that original string, by
+  // just calling the charCodeAt API on it.
+  let utf8Str : string;
+  const pcStr = encodeURIComponent(str);
+
+  // As "unescape" is a deprecated function we want to declare a fallback in the
+  // case a browser decide to not implement it.
+  if (typeof window.unescape === "function") {
```

Comment:
Don't forget to check if is node

Reply:
Yes, maybe even removing `window` altogether:
  - with the `typeof` operator, I think that you can use undefined variables
  - `unescape` actually exists in nodejs

Reply:
Done